allennlp/models/encoder_decoders/nlvr_semantic_parser.py

```diff
@@ -479,13 +551,123 @@ def __init__(self,
         self.action_embeddings = action_embeddings
         self.action_indices = action_indices
         self.possible_actions = possible_actions
+        self.worlds = worlds
+        self.label_strings = label_strings
 
-    def get_valid_actions(self) -> List[List[int]]:
+    def get_valid_actions(self, shuffle: bool = True) -> List[List[int]]:
         """
         This method is identical to ``WikiTablesDecoderState.get_valid_actions``.
         Returns a list of valid actions for each element of the group.
         """
-        return [state.get_valid_actions() for state in self.grammar_state]
+        valid_actions = [state.get_valid_actions() for state in self.grammar_state]
+        if shuffle:
```

Comment:
When we have a fixed order for iterating over valid actions, the model is more likely to see the earlier actions than the later ones because there's a beam, especially in the case of non-agenda actions. I suspected that this may be biasing the model towards them. But shuffling did not result in noticeable changes in the accuracy.

Reply:
If you have a model score, not just a coverage score, this isn't true - the actions will get re-sorted by model score, which is initialized basically randomly.  There's no way that you'd end up with the same ordering on the beam as you saw here.